### PR TITLE
Create skeleton structure for "I/O Optimisations" sections

### PR DESCRIPTION
**Pull request by teodutu**

### CODE REVIEW

**Comment by Alex-deVis** [comment]
I think we can add `I/O buffering` as a section.

**Comment by teodutu** [comment]
@Alex-deVis "I/O Buffering" will be part of the "I/O Internals section". I plan to first show that `struct _IO_FILE` defines internal buffers and then explain why. I think it's easier to start with a real-life example in something like Musl (and a bit of glibc) so students get the feeling that this is an important optimisation from the beginning.

File: content/chapters/io/lab/content/io-internals.md [hunk 1]
```diff
@@ -45,6 +45,28 @@ struct _IO_FILE {
 };
 ```
 
+## IO Optimisations
```
Comment (Alex-deVis) [inline content/chapters/io/lab/content/io-internals.md:48]:
```suggestion
## IO Optimizations
```

File: content/chapters/io/lab/quiz/mmap-read-write-benchmark.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+- the one using `read()` and `write()`
+
++ they are roughly equivalent
+
+## Feedback
+
+In pur case, running the script a few times results in the following running times:
+
+```console
+teo@lautari file-mappings $ ./benchmark_cp.sh
```
Comment (Alex-deVis) [inline content/chapters/io/lab/quiz/mmap-read-write-benchmark.md:20]:
```suggestion
student@OS file-mappings $ ./benchmark_cp.sh 
```
Even though it is not as cool.

File: content/chapters/io/lab/quiz/mmap-read-write-benchmark.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+user    0m0,569s
+sys	0m2,286s
+Benchmarking cp on a 1 GB file...
+
+real    0m36,012s
+user    0m0,039s
+sys	0m2,469s
+
+
+teo@lautari file-mappings $ ./benchmark_cp.sh
```
Comment (Alex-deVis) [inline content/chapters/io/lab/quiz/mmap-read-write-benchmark.md:34]: ditto.

File: content/chapters/io/lab/content/io-internals.md [hunk 4]
```diff
@@ -45,6 +45,28 @@ struct _IO_FILE {
...
+
+On the other hand, as we've already established, the I/O component defines how we interact with an app.
+If we want it to be responsive and to do something useful, most likely, the I/O is the key.
+
+So I/O is crucial for most applications, yet it is also the slowest...
+
+![Sad Pepe](../media/sad-pepe.png)
+
+But fear not!
+There are countless optimisations out there aimed precisely at bridging the speed gap between the memory and the disk.
```
Comment (Alex-deVis) [inline content/chapters/io/lab/content/io-internals.md:68]:
```suggestion
There are countless optimizations out there aimed precisely at bridging the speed gap between the memory and the disk.
```

File: content/chapters/io/lab/support/async/python/async_server_3.6.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    server = loop.run_until_complete(coro)
+
+    try:
+        loop.run_forever()
+    except KeyboardInterrupt:
+        pass
+
+    server.close()
+    loop.run_until_complete(server.wait_closed())
+    loo.close()
```
Comment (Alex-deVis) [inline content/chapters/io/lab/support/async/python/async_server_3.6.py:52]:
```suggestion
    loop.close()
```

File: content/chapters/io/lab/content/file-mappings.md [hunk 6]
```diff
@@ -1 +1,105 @@
...
+
+## The Return of `mmap()`
+
+Remember that the `.text`, `.rodata` and `.data` sections of libraries are present in the VAS of any Linux process:
+
+```console
+student@os:~$ sleep 1000 &  # start a `sleep` process in the background
+[1] 17579
+
+student@os:~$ sleep 1000 cat /proc/$(pidof sleep)/maps
```
Comment (gabrielmocanu) [inline content/chapters/io/lab/content/file-mappings.md:15]:
```suggestion
student@os:~$ cat /proc/$(pidof sleep)/maps
```

**Review by gabrielmocanu — CHANGES_REQUESTED** [review]

File: content/chapters/io/lab/support/async/python/async_server.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+
+# https://docs.python.org/3/library/asyncio-stream.html
+
+import sys
+import asyncio
+import socket
```
Comment (Alex-deVis) [inline content/chapters/io/lab/support/async/python/async_server.py:7]:
```suggestion

```

File: content/chapters/io/lab/support/async/python/mp_server.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    connection.close()
+
+
+LISTEN_BACKLOG = 10
+
+def run_server(port, hostname="0.0.0.0"):
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.bind((hostname, port))
+    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
```
Comment (Alex-deVis) [inline content/chapters/io/lab/support/async/python/mp_server.py:38]:
```suggestion
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((hostname, port))
```
`setsockopt()` must be called before `bind()`.

File: content/chapters/io/lab/support/async/python/mt_server.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    connection.close()
+
+
+LISTEN_BACKLOG = 10
+
+def run_server(port, hostname="0.0.0.0"):
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.bind((hostname, port))
+    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
```
Comment (Alex-deVis) [inline content/chapters/io/lab/support/async/python/mt_server.py:38]:
```suggestion
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((hostname, port))
```
`setsockopt()` must be called before `bind()`.

File: content/chapters/io/lab/support/async/python/server.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+    connection.close()
+
+
+LISTEN_BACKLOG = 10
+
+def run_server(port, hostname="0.0.0.0"):
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.bind((hostname, port))
+    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
```
Comment (Alex-deVis) [inline content/chapters/io/lab/support/async/python/server.py:37]:
```suggestion
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((hostname, port))
```
`setsockopt()` must be called before `bind()`.

File: content/chapters/io/lab/support/async/python/client_bench.sh [hunk 11]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+NUM_CLIENTS=8
+
+for i in $(seq 1 $NUM_CLIENTS); do
+    ./client.py localhost 2999 30 &
+done
```
Comment (Alex-deVis) [inline content/chapters/io/lab/support/async/python/client_bench.sh:7]:
```suggestion
if [[ $# != 1 ]]; then
    echo "usage: $0 <port>"
    exit 1
fi

NUM_CLIENTS=8
PORT=$1

for i in $(seq 1 $NUM_CLIENTS); do
    ./client.py localhost "$1" 30 &
done
```
Take `PORT` as command line argument.

Comment (teodutu) [reply content/chapters/io/lab/support/async/python/client_bench.sh:7]: Not really needed. Port 2990 is not that common. Maybe if it were 6969...

**Review by Alex-deVis — APPROVED** [review]
Looks good.

**Review by razvand — APPROVED** [review]
Works

**Pull request merged**